inlong-sort/sort-connectors/tubemq/src/main/java/org/apache/inlong/sort/tubemq/FlinkTubeMQConsumer.java

```diff
@@ -0,0 +1,362 @@
+                new ListStateDescriptor<>(TUBE_OFFSET_STATE, typeInformation);
+
+        OperatorStateStore stateStore = context.getOperatorStateStore();
+        offsetsState = stateStore.getListState(stateDescriptor);
+
+        currentOffsets = new HashMap<>();
+        if (context.isRestored()) {
+            for (Tuple2<String, Long> tubeOffset : offsetsState.get()) {
+                currentOffsets.put(tubeOffset.f0, tubeOffset.f1);
+            }
+
+            LOG.info("Successfully restore the offsets {}.", currentOffsets);
+        } else {
+            LOG.info("No restore offsets.");
+        }
+    }
+
+    @Override
+    public void open(Configuration parameters) throws Exception {
+        ConsumerConfig consumerConfig =
+                new ConsumerConfig(masterAddress, consumerGroup);
+        consumerConfig.setConsumePosition(consumeFromMax
+                ? ConsumePosition.CONSUMER_FROM_MAX_OFFSET_ALWAYS
+                : ConsumePosition.CONSUMER_FROM_FIRST_OFFSET);
+
+        consumerConfig.setMsgNotFoundWaitPeriodMs(messageNotFoundWaitPeriod.toMillis());
+
+        final int numTasks = getRuntimeContext().getNumberOfParallelSubtasks();
+
+        messageSessionFactory = new TubeSingleSessionFactory(consumerConfig);
+
+        messagePullConsumer =
+                messageSessionFactory.createPullConsumer(consumerConfig);
+        messagePullConsumer
+                .subscribe(topic, tidSet);
+        messagePullConsumer
+                .completeSubscribe(sessionKey, numTasks, true, currentOffsets);
+
+        running = true;
+    }
+
+    @Override
+    public void run(SourceContext<T> ctx) throws Exception {
+
+        Instant lastConsumeInstant = Instant.now();
+
+        while (running) {
+
+            ConsumerResult consumeResult = messagePullConsumer.getMessage();
+            if (!consumeResult.isSuccess()) {
+                if (!(consumeResult.getErrCode() == 400
+                        || consumeResult.getErrCode() == 404
+                        || consumeResult.getErrCode() == 405
+                        || consumeResult.getErrCode() == 406
+                        || consumeResult.getErrCode() == 407
+                        || consumeResult.getErrCode() == 408)) {
+                    LOG.info("Could not consume messages from tubemq (errcode: {}, "
+                                    + "errmsg: {}).", consumeResult.getErrCode(),
+                            consumeResult.getErrMsg());
+                }
+
+                Duration idleTime =
+                        Duration.between(lastConsumeInstant, Instant.now());
+                if (idleTime.compareTo(maxIdleTime) > 0) {
+                    ctx.markAsTemporarilyIdle();
+                }
+
+                continue;
+            }
+
+            List<Message> messageList = consumeResult.getMessageList();
+            lastConsumeInstant = Instant.now();
+
+            List<T> records = new ArrayList<>();
+            if (messageList != null) {
```

Comment:
whether the case  of  messageList.length is 0  is  correct?

Reply:
No this case